test/test_torch.py

```diff
@@ -3889,6 +3889,44 @@ def test_comparison_ops_must_take_bool_output(self):
                        torch.logical_and, torch.logical_or, torch.logical_xor]:
                 self.assertEqual(op(torch.tensor([True]), torch.tensor([False])).dtype, torch.bool)
 
+        def test_out_comparison_ops_type_promotion_and_broadcasting(self):
```

Comment:
I think these tests can expand their coverage while being simplified.

The @dtypes decorator lets you specify dtype pairs. The tests can enumerate test cases with different sizes, call make_tensor with the size and the appropriate dtype, and compare the results of each of these ops with NumPy.  The out variant promotion can be tested by using out=complex128, which every type is safe-castable to. 

This would cover most cases but not extremal cases that require fixtures. So after the shape-based test cases a value-based enumeration of test cases can be added. In pseucode:

- for every op
- for every shape x shape test case
- create tensor in dtype0 with shape0, create tensor in dtype1 with shape1
- run the op and compare with NumPy
- run the op with out=complex128 and compare with NumPy
- for every value x value test case... (repeat the above)